### DISCUSSION TITLE
Backstepping control.

### DISCUSSION THREAD

**Discussion by Carmol1** [bitcraze/discussions#185] [Q&A] [answered]
Hello.

Is it possible to make use of the controllers published in this paper  "Backstepping and Sliding-mode Techniques Applied to an Indoor Micro Quadrotor" in the crazyflie?

I'd appreciate your help

Regards.

**Comment by knmcguire** [comment, marked as answer]
Anything is possible since the firmware is opensource. I haven't read the paper myself, but it is published before the crazyflie existed so it has not been implemented in the firmware yet. Also I'm not aware of any offboard implementations of this approach either so that is something you will need to investigate.

On control itself, it really depends what you want to do... offboard control or onboard control. 
- Offboard control is easier to implement and you could use our [python library](https://github.com/bitcraze/crazyflie-lib-python) for that but there are going to be communication delays, which makes it impossible to go all the way down to attitude rate control from an external computer.
- Onboard control means very little delay from commands to motors, but more difficult to implement on the [c-based firmware](https://github.com/bitcraze/crazyflie-firmware). [This gives an overview ](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/)of which controllers we have on the crazyflie. For implementation, you can take a look on how others has done it like for the more recent INDI controller in [this PR](https://github.com/bitcraze/crazyflie-firmware/pull/483) and [this one ](https://github.com/bitcraze/crazyflie-firmware/pull/568).

Hope this helps!

**Comment by Carmol1** [comment]
Thank you for your response, finally, if I want to program it onboard, is it enough to modify an existing file (Mellinger, PID, INDI)? or is it necessary to start from scratch?.

Regards.

**Reply by knmcguire** [reply]
That is probably the easiest to do it, as long as you don't have ambition to recontribute it back to the original source code. 

If that is your plan to make it part of the official crazyflie firmware in the future, the it is better to make a new controller in the style of the others. The same holds if you want to easily switch between controllers for comparison reasons.

**Reply by Carmol1** [reply]
Thank you so much for your help.

**Comment by knmcguire** [comment]
Just came across this publication that might be of interest to you @Carmol1 https://www.sciencedirect.com/science/article/abs/pii/S096706612200140X
